test/e2e/olm_test.go

```diff
@@ -0,0 +1,218 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	guestClusterCatalogSourceName    = "guestcluster-operators"
+	redhatOperatorsCatalogSourceName = "redhat-operators"
+	redhatOperatorsDeploymentName    = "redhat-operators-catalog"
+	openshiftMarketplaceNamespace    = "openshift-marketplace"
+)
+
+// TestOLM executes a suite of olm tests which ensure that olm is
+// behaving as expected on the guest cluster.
+func TestOLM(t *testing.T) {
+	t.Parallel()
+
+	g := NewWithT(t)
+
+	ctx, cancel := context.WithCancel(testContext)
+	defer cancel()
+
+	client := e2eutil.GetClientOrDie()
+
+	// Create a cluster
+	clusterOpts := globalOpts.DefaultClusterOptions()
+	clusterOpts.NodePoolReplicas = 1
+	cluster := e2eutil.CreateCluster(t, ctx, client, clusterOpts, globalOpts.ArtifactDir)
+
+	// Get guest client
+	t.Logf("Waiting for guest client to become available")
+	guestClient := e2eutil.WaitForGuestClient(t, ctx, client, cluster)
+
+	// Wait for guest cluster nodes to become available
+	util.WaitForNReadyNodes(t, ctx, guestClient, clusterOpts.NodePoolReplicas)
+
+	guestNamespace := manifests.HostedControlPlaneNamespace(cluster.Namespace, cluster.Name).Name
+	t.Logf("Hosted control plane namespace is %s", guestNamespace)
+
+	t.Logf("Retrieving Red Hat catalogSource deployment image")
+	redhatCatalogDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      redhatOperatorsDeploymentName,
+			Namespace: guestNamespace,
+		},
+	}
+	err := wait.PollImmediateUntil(5*time.Second, func() (bool, error) {
+		if err := client.Get(ctx, crclient.ObjectKeyFromObject(redhatCatalogDeployment), redhatCatalogDeployment); err != nil {
+			t.Logf("failed to get Red Hat Catalog deployment %s/%s: %s", redhatCatalogDeployment.Namespace, redhatCatalogDeployment.Name, err)
+			return false, nil
+		}
+		return true, nil
+	}, ctx.Done())
+
+	t.Logf("Creating guest cluster catalogSource using the Red Hat operators image from the hosted control plane")
+	guestClusterCatalogSource := &operatorsv1alpha1.CatalogSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      guestClusterCatalogSourceName,
+			Namespace: openshiftMarketplaceNamespace,
+		},
+		Spec: operatorsv1alpha1.CatalogSourceSpec{
+			SourceType:  operatorsv1alpha1.SourceTypeGrpc,
+			Image:       redhatCatalogDeployment.Spec.Template.Spec.Containers[0].Image,
+			Publisher:   "OLM HyperShift Team",
+			DisplayName: "OLM HyperShift Test CatalogSource",
+		},
+	}
+	err = wait.PollImmediateUntil(5*time.Second, func() (bool, error) {
+		if err := guestClient.Create(ctx, guestClusterCatalogSource); err != nil && !apierrors.IsAlreadyExists(err) {
+			t.Logf("failed to create guest cluster catalogSource %s/%s: %s", guestClusterCatalogSource.Namespace, guestClusterCatalogSource.Name, err)
+			return false, nil
+		}
+		return true, nil
```

Comment:
Just a quick thought: do we need to also wait for "readiness" to avoid any race conditions when creating subscriptions in later sub-tests? I feel like this should be fine just wanted to check whether it's needed to avoid introducing any potential flakiness.

Reply:
The subscription references the catalog source that already reported "ready". Could be done for the RedHat catalog though, but that is typically resolved before the guest cluster worker nodes are created and running.